### PR TITLE
Support Coverity Scan

### PR DESCRIPTION
**Pull request by bennylp**
Work in progress.

See [tests/coverity-scan/README.md](https://github.com/pjsip/pjproject/blob/coverity01/tests/coverity-scan/README.md) for instructions

https://scan.coverity.com/projects/pjsip
Initial outstanding: 416, high impact: 67, fixed: 155
Host: fresh Ubuntu Docker (installed with `run.sh`)

### CODE REVIEW

**Comment by bennylp** [comment]
Outstanding: 498 (52 high impact).
Maybe outstanding is increased because I'm submitting from different host?

**Comment by bennylp** [comment]
Outstanding: 463, high impact: 17, fixed: 203
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
Outstanding: 420, high impact: 12, fixed: 203
Host: Ubuntu Docker (installed with run.sh)

**Comment by sauwming** [comment]
There are ~127 reports classified as medium with type `Invalid type in argument to printf format specifier`. I have fixed some, but I'm not sure if this is important and perhaps can be filtered out?

**Comment by bennylp** [comment]
Outstanding: 281, high impact: 1, dismissed: 164, fixed: 247
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
Outstanding: 271, high impact: 1, dismissed: 165, fixed: 256
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
Outstanding: 206, high impact: 3, dismissed: 189, fixed: 297
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
Outstanding: 182, high impact: 0, dismissed: 202, fixed: 297
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
For defects which I think are legitimate and I'm unable to fix, I set the Action to `Fix Required`.
You can view these using "*Confirmed Defects*" filter.

IMO these non-trivial defects should be fixed in separate/individual issues?

**Comment by jimying** [comment]
how about replace pj_ansi_strncpy() implement  by bsd strlcpy() ?
https://github.com/freedesktop/libbsd/blob/master/src/strlcpy.c

**Comment by bennylp** [comment]
Outstanding: 147, dismissed: 217, fixed: 324
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
> There are ~127 reports classified as medium with type `Invalid type in argument to printf format specifier`. I have fixed some, but I'm not sure if this is important and perhaps can be filtered out?

IMO they should be fixed (either by typecasting the argument or by changing the format) as they are genuine warnings. If the compiler knew that PJ_LOG takes printf style arguments, it will generate the same warnings.

To prevent the same problem in the future, perhaps we can tell the compiler that pj_log_xx takes printf style arguments, at least for gcc:

```
log.h:

#ifdef __GNUC__
# define PJ_LOG_FUNC_DECOR __attribute__((format (printf, 2, 3)))
#else
# define PJ_LOG_FUNC_DECOR
#endif

...
    PJ_DECL(void) pj_log_1(const char *src, const char *format, ...)
                  PJ_LOG_FUNC_DECOR;
...
   // and so on for pj_log_2 until 6
```

References:
- https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Function-Attributes.html
- Unfortunately this feature is not implemented in Visual Studio yet: https://developercommunity.visualstudio.com/t/support-format-specifier-checking-on-user-defined/843652

**Comment by sauwming** [comment]
> To prevent the same problem in the future, perhaps we can tell the compiler that pj_log_xx takes printf style arguments, at least for gcc:

Even though it doesn't work on all platforms, I think it's a good idea.
I tried it and it also works on clang on Mac.

At least this way, we can quickly fix it locally, rather than having to wait for the submission result.

**Comment by bennylp** [comment]
Re: the snippet, I think we should move the decoration declaration to somewhere else more general (e.g. `pj/types.h`), as I found two other occurrences of printf style functions:

- `pj_perror[_xx]`
- `pjsip_endpt_log_error`

**Comment by sauwming** [comment]
> i think it better revert pj_ansi_strncpy to original
> ```c
> #define pj_ansi_strncpy         strncpy
> ```
> and add new safe method pj_ansi_strlcpy( ) like bsd os.

Initially we did the same thing (we implemented `pj_ansi_safe_strncpy()`):
https://github.com/pjsip/pjproject/pull/3390/commits/5fe4bc1972f9475c84956969dfdcd1ac058ff224 (`Fix usages of strncpy(), pj_ansi_strncpy(), and strncat() by replacing them with pj_ansi_safe_strncpy() and pj_ansi_safe_strcpycat()`)

Then we agreed to make all `pj_ansi_strncpy()` safe:
https://github.com/pjsip/pjproject/pull/3390/commits/b97614d1bc7f8e7a7d2f952e6a2db48c7e1d4f1d (`Drop pj_ansi_safe_strncpy() in favor of safe pj_ansi_strncpy()`)

Little did we know that modifying `pj_ansi_strncpy()` may break applications, as evidenced by a couple of our own usages within the library. So it's not impossible it may cause user's applications to break, too.

So I guess we need to go back to the first approach again, i.e. using `pj_ansi_safe_strncpy()` or `pj_ansi_strlcpy()`?

File: pjlib/include/pj/errno.h
Comment (jimying) [inline pjlib/include/pj/errno.h:1]: @sauwming This line is missing  the printf format check macro PJ_PRINT_FUNC_DECOR

Comment (sauwming) [reply pjlib/include/pj/errno.h:1]: thanks, you are a lifesaver.

**Comment by sauwming** [comment]
> For defects which I think are legitimate and I'm unable to fix, I set the Action to `Fix Required`. You can view these using "_Confirmed Defects_" filter.
> 
> IMO these non-trivial defects should be fixed in separate/individual issues?

Done. From the 6, I created two PRs, and 4 separate commits.

**Comment by bennylp** [comment]
> So it's not impossible it may cause user's applications to break, too.

True. 

> So I guess we need to go back to the first approach again, i.e. using `pj_ansi_safe_strncpy()` or `pj_ansi_strlcpy()`?

I think `pj_ansi_strscpy()` is probably what we want. P.S. there is an interesting article on this issue: https://mafford.com/text/the-many-ways-to-copy-a-string-in-c/

**Comment by jimying** [comment]
> > So it's not impossible it may cause user's applications to break, too.
> 
> True.
> 
> > So I guess we need to go back to the first approach again, i.e. using `pj_ansi_safe_strncpy()` or `pj_ansi_strlcpy()`?
> 
> I think `pj_ansi_strscpy()` is probably what we want. P.S. there is an interesting article on this issue: https://mafford.com/text/the-many-ways-to-copy-a-string-in-c/


The Git use strlcpy ^_^ 
https://github.com/git/git/blob/27d43aaaf50ef0ae014b88bba294f93658016a2e/git-compat-util.h#L890

**Comment by jimying** [comment]
One print format warning
```
n file included from ../src/pj/ssl_sock_ossl.c:26:
../src/pj/ssl_sock_ossl.c: In function ‘ssl_do_handshake’:
../src/pj/ssl_sock_ossl.c:2367:35: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘long int’ [-Wformat=]
 2367 |             PJ_LOG(5, (THIS_FILE, "Session info: reused=%d, resumable=%d, "
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../include/pj/log.h:460:50: note: in definition of macro ‘pj_log_wrapper_5’
  460 |     #define pj_log_wrapper_5(arg)       pj_log_5 arg 
      |                                                  ^~~
../src/pj/ssl_sock_ossl.c:2367:13: note: in expansion of macro ‘PJ_LOG’
 2367 |             PJ_LOG(5, (THIS_FILE, "Session info: reused=%d, resumable=%d, "
      |             ^~~~~~
../src/pj/ssl_sock_ossl.c:2367:58: note: format string is defined here
 2367 |             PJ_LOG(5, (THIS_FILE, "Session info: reused=%d, resumable=%d, "
      |                                                         ~^
      |                                                          |
      |                                                          int
      |                                                         %ld
```

**Comment by jimying** [comment]
two missing print format warnings:
```
../src/pjsip/sip_endpoint.c:1335:25: warning: format specifies type 'unsigned int' but the argument has type 'pj_size_t' (aka 'unsigned long') [-Wformat]
                           pj_timer_heap_count(endpt->timer_heap)));
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
../src/pjmedia-codec/vid_toolbox.m:1374:19: warning: format specifies type 'int' but the argument has type 'pj_size_t' (aka 'unsigned long') [-Wformat]
                  count, whole_len));
                  ^~~~~
```

**Comment by bennylp** [comment]
Outstanding: 138, dismissed: 97, fixed: 530, new: 0
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
Coverity scan queued (this comment will be edited once it finishes)

**Comment by bennylp** [comment]
Reserved

**Comment by bennylp** [comment]
Change: un-ignore third_party but dismiss the defects

Outstanding: 0, dismissed: 192, fixed: 573, new: 0
Host: Ubuntu Docker (installed with run.sh)

**Comment by bennylp** [comment]
In https://github.com/pjsip/pjproject/pull/3390/commits/5ed9461bbe6a8a81c63fe1acc7feb353ba2f3980 and https://github.com/pjsip/pjproject/pull/3390/commits/7a058dee8108924011fb8e5800c57b829bb21a29:

- Implement `pj_ansi_strxcpy()`, `pj_ansi_strxcpy2()`, and `pj_ansi_strxcat()`
- Replace all unsafe string functions with their safer counterpart:

  - `strcpy/pj_ansi_strcpy    --> pj_ansi_strxcpy`
  - `strncpy/pj_ansi_strncpy  --> pj_ansi_strxcpy`
  - `strcat/pj_ansi_strcat    --> pj_ansi_strxcat`
  - `sprintf/pj_ansi_sprintf  --> pj_ansi_snprintf`
  - `vsprintf/pj_ansi_vsprintf  --> pj_ansi_vsnprintf`

- Introduce new macro, `PJ_BAN_UNSAFE_STR_FUNCS`, if set to 1, will prohibit the use of unsafe string functions above. By default it is set to 0. Feel free to declare this to 1 in your `config_site.h`. Or perhaps should we set this in makefiles when building PJ libs?
- TODO: create safer variant for unsafe pj_str_t functions such as `pj_strcat`.

**Review by sauwming — APPROVED** [review]

**Comment by jimying** [comment]
hi,  I had usedclang scan-build (#3443 ) to check this branch code. There are still some warning reports.  Better to fix them by the way?


```bash
JOBS=8
SCAN_BUILD="scan-build -o /tmp/pjsip"
export CC=clang
export CXX=clang++
export CFLAGS="-g -fPIC"

make distclean

./configure .....

make dep

# one sperate sub directory per module
$SCAN_BUILD make -C pjlib/build -j  $JOBS
$SCAN_BUILD make -C pjlib-util/build -j  $JOBS
$SCAN_BUILD make -C pjnath/build -j  $JOBS
$SCAN_BUILD make -C third_party/build -j  $JOBS
$SCAN_BUILD make -C pjmedia/build -j  $JOBS
$SCAN_BUILD make -C pjsip/build -j $JOBS
$SCAN_BUILD make -C pjsip-apps/build -j $JOBS

```

**Comment by sauwming** [comment]
Since this PR is already large enough, I think we'll work with `scan-build` on a separate PR.

**Review by nanangizz — APPROVED** [review]

File: pjlib/include/pj/string.h [hunk 1]
```diff
@@ -840,6 +840,58 @@ PJ_INLINE(void*) pj_memchr(const void *buf, int c, pj_size_t size)
...
+/**
+ * Concatenate src, or as much of it as fits, into the dest buffer.
+ * Regardless of whether all characters were copied, the destination
+ * buffer will be null terminated, unless dst_size is zero which in
+ * this case nothing will be written to dst and the function will
+ * return -PJ_ETOOBIG.
+ *
+ * @param dst       The destination string.
+ * @param src       The source string.
+ * @param dst_size  The size of the destination string.
```
Comment (jimying) [inline pjlib/include/pj/string.h:887]: Better to add comment for dst_size:
"the dst_size is full size of dst, not space left"

File: pjlib/src/pj/addr_resolv_sock.c [hunk 2]
```diff
@@ -220,11 +221,11 @@ PJ_DEF(pj_status_t) pj_getaddrinfo(int af, const pj_str_t *nodename,
 
         /* Store canonical name (possibly truncating the name) */
         if (res->ai_canonname) {
-            pj_ansi_strncpy(ai[i].ai_canonname, res->ai_canonname,
-                            sizeof(ai[i].ai_canonname));
-            ai[i].ai_canonname[sizeof(ai[i].ai_canonname)-1] = '\0';
+            pj_ansi_strxcpy(ai[i].ai_canonname, res->ai_canonname,
+                                 sizeof(ai[i].ai_canonname));
```
Comment (jimying) [inline pjlib/src/pj/addr_resolv_sock.c:225]: a code indent problem

File: pjlib/src/pj/addr_resolv_sock.c [hunk 3]
```diff
@@ -303,9 +304,8 @@ PJ_DEF(pj_status_t) pj_getaddrinfo(int af, const pj_str_t *nodename,
         pj_bzero(ai, max_count * sizeof(pj_addrinfo));
 
         for (i=0; he.h_addr_list[i] && *count<max_count; ++i) {
-            pj_ansi_strncpy(ai[*count].ai_canonname, he.h_name,
-                            sizeof(ai[*count].ai_canonname));
-            ai[*count].ai_canonname[sizeof(ai[*count].ai_canonname)-1] = '\0';
+            pj_ansi_strxcpy(ai[*count].ai_canonname, he.h_name,
+                                 sizeof(ai[*count].ai_canonname));
```
Comment (jimying) [inline pjlib/src/pj/addr_resolv_sock.c:308]: a code indent problem

File: pjlib/src/pj/os_info.c [hunk 4]
```diff
@@ -70,26 +70,26 @@
...
-
     if (ver == 0) {
         *buf = '\0';
         return buf;
     }
 
-    sprintf(buf, "-%u.%u",
+    pj_ansi_snprintf(buf, buf_size, "-%u.%u",
             (ver & 0xFF000000) >> 24,
             (ver & 0x00FF0000) >> 16);
```
Comment (jimying) [inline pjlib/src/pj/os_info.c:82]: a code indent problem

File: pjlib-util/src/pjlib-util-test/http_client.c [hunk 5]
```diff
@@ -116,14 +116,24 @@ static int server_thread(void *p)
...
+            PJ_ASSERT_ON_FAIL(pkt_len>0, {
+                PJ_PERROR(2, (THIS_FILE, -pkt_len, "Error creating response"));
+                pj_sock_close(newsock);
+                continue;
+            })
             if (srv->send_content_length) {
-                pj_ansi_sprintf(pkt + pj_ansi_strlen(pkt), 
+                pj_ansi_snprintf(pkt + pkt_len, srv->buf_size - pkt_len,
                                 "Content-Length: %d\r\n",
                                 srv->data_size);
```
Comment (jimying) [inline pjlib-util/src/pjlib-util-test/http_client.c:129]: a code indent  problem.  Better to check all `sprintf`  change to `snprintf`

File: pjmedia/src/pjmedia-audiodev/wmme_dev.c [hunk 6]
```diff
@@ -330,20 +330,19 @@ static void build_dev_info(UINT deviceId, struct wmme_dev_info *wdi,
 
     /* Device Name */
     if (deviceId==WAVE_MAPPER) {
-        strncpy(wdi->info.name, "Wave mapper", sizeof(wdi->info.name));
-        wdi->info.name[sizeof(wdi->info.name)-1] = '\0';
+        pj_ansi_strxcpy(wdi->info.name, "Wave mapper", 
+                             sizeof(wdi->info.name));
```
Comment (jimying) [inline pjmedia/src/pjmedia-audiodev/wmme_dev.c:334]: a code indent problem

File: pjmedia/src/pjmedia-audiodev/wmme_dev.c [hunk 7]
```diff
@@ -330,20 +330,19 @@ static void build_dev_info(UINT deviceId, struct wmme_dev_info *wdi,
...
     } else {
         const pj_char_t *szPname = WIC_WOC(wic, woc, szPname);
         PJ_DECL_ANSI_TEMP_BUF(wTmp, sizeof(wdi->info.name));
         
-        strncpy(wdi->info.name, 
+        pj_ansi_strxcpy(wdi->info.name, 
                 PJ_NATIVE_TO_STRING(szPname, wTmp, PJ_ARRAY_SIZE(wTmp)),
                 sizeof(wdi->info.name));
```
Comment (jimying) [inline pjmedia/src/pjmedia-audiodev/wmme_dev.c:341]: a code indent problem

File: pjmedia/src/pjmedia-videodev/avi_dev.c [hunk 8]
```diff
@@ -281,8 +281,9 @@ static void reset_dev_info(struct avi_dev_info *adi)
     pj_bzero(adi, sizeof(*adi));
 
     /* Fill up with *dummy" device info */
-    pj_ansi_strncpy(adi->info.name, "AVI Player", sizeof(adi->info.name)-1);
-    pj_ansi_strncpy(adi->info.driver, DRIVER_NAME, sizeof(adi->info.driver)-1);
+    pj_ansi_strxcpy(adi->info.name, "AVI Player", sizeof(adi->info.name));
+    pj_ansi_strxcpy(adi->info.driver, DRIVER_NAME, 
+                         sizeof(adi->info.driver));
```
Comment (jimying) [inline pjmedia/src/pjmedia-videodev/avi_dev.c:286]: a code indent problem

File: pjmedia/src/pjmedia-videodev/avi_dev.c [hunk 9]
```diff
@@ -484,8 +488,10 @@ PJ_DEF(pj_status_t) pjmedia_avi_dev_alloc( pjmedia_vid_dev_factory *f,
     }
 
     /* Init device info */
-    pj_ansi_strncpy(adi->info.name, adi->title.ptr, sizeof(adi->info.name)-1);
-    pj_ansi_strncpy(adi->info.driver, DRIVER_NAME, sizeof(adi->info.driver)-1);
+    pj_ansi_strxcpy(adi->info.name, adi->title.ptr, 
+                         sizeof(adi->info.name));
+    pj_ansi_strxcpy(adi->info.driver, DRIVER_NAME, 
+                         sizeof(adi->info.driver));
```
Comment (jimying) [inline pjmedia/src/pjmedia-videodev/avi_dev.c:494]: a code indent problem

File: pjmedia/src/pjmedia-videodev/colorbar_dev.c [hunk 10]
```diff
@@ -201,11 +201,10 @@ static pj_status_t cbar_factory_init(pjmedia_vid_dev_factory *f)
...
     pj_bzero(ddi, sizeof(*ddi));
-    pj_ansi_strncpy(ddi->info.name, "Colorbar generator",
-                    sizeof(ddi->info.name));
-    ddi->info.driver[sizeof(ddi->info.driver)-1] = '\0';
-    pj_ansi_strncpy(ddi->info.driver, "Colorbar", sizeof(ddi->info.driver));
-    ddi->info.driver[sizeof(ddi->info.driver)-1] = '\0';
+    pj_ansi_strxcpy(ddi->info.name, "Colorbar generator",
+                         sizeof(ddi->info.name));
+    pj_ansi_strxcpy(ddi->info.driver, "Colorbar", 
+                         sizeof(ddi->info.driver));
```
Comment (jimying) [inline pjmedia/src/pjmedia-videodev/colorbar_dev.c:207]: a code indent problem

File: pjmedia/src/pjmedia-videodev/darwin_dev.m [hunk 11]
```diff
@@ -361,9 +361,11 @@ static pj_status_t darwin_factory_refresh(pjmedia_vid_dev_factory *f)
...
             qdi = &qf->dev_info[qf->dev_count++];
             pj_bzero(qdi, sizeof(*qdi));
-            pj_ansi_strncpy(qdi->info.name, [device.localizedName UTF8String],
-                            sizeof(qdi->info.name));
-            pj_ansi_strncpy(qdi->info.driver, "AVF", sizeof(qdi->info.driver));
+            pj_ansi_strxcpy(qdi->info.name, 
+                                 [device.localizedName UTF8String],
+                                 sizeof(qdi->info.name));
+            pj_ansi_strxcpy(qdi->info.driver, "AVF", 
+                                 sizeof(qdi->info.driver));
```
Comment (jimying) [inline pjmedia/src/pjmedia-videodev/darwin_dev.m:368]: a code indent problem

File: pjmedia/src/pjmedia-videodev/sdl_dev.c [hunk 12]
```diff
@@ -493,22 +493,22 @@ static pj_status_t sdl_factory_init(pjmedia_vid_dev_factory *f)
...
     ddi->info.fmt_cnt = 1;
 #endif /* PJMEDIA_VIDEO_DEV_SDL_HAS_OPENGL */
 
     for (i = 0; i < sf->dev_count; i++) {
         ddi = &sf->dev_info[i];
-        strncpy(ddi->info.driver, "SDL", sizeof(ddi->info.driver));
-        ddi->info.driver[sizeof(ddi->info.driver)-1] = '\0';
+        pj_ansi_strxcpy(ddi->info.driver, "SDL", 
+                             sizeof(ddi->info.driver));
         ddi->info.dir = PJMEDIA_DIR_RENDER;
```
Comment (jimying) [inline pjmedia/src/pjmedia-videodev/sdl_dev.c:512]: code indent problem

File: pjmedia/src/pjmedia/transport_adapter_sample.c [hunk 13]
```diff
@@ -125,8 +125,8 @@ PJ_DEF(pj_status_t) pjmedia_tp_adapter_create( pjmedia_endpt *endpt,
     pool = pjmedia_endpt_create_pool(endpt, name, 512, 512);
     adapter = PJ_POOL_ZALLOC_T(pool, struct tp_adapter);
     adapter->pool = pool;
-    pj_ansi_strncpy(adapter->base.name, pool->obj_name, 
-                    sizeof(adapter->base.name));
+    pj_ansi_strxcpy(adapter->base.name, pool->obj_name, 
+                         sizeof(adapter->base.name));
```
Comment (jimying) [inline pjmedia/src/pjmedia/transport_adapter_sample.c:129]: a code indent problem

File: pjsip-apps/src/pjsua/pjsua_app_config.c [hunk 14]
```diff
@@ -1601,7 +1601,7 @@ static void default_config()
     pjsua_app_config *cfg = &app_config;
 
     pjsua_config_default(&cfg->cfg);
-    pj_ansi_sprintf(tmp, "PJSUA v%s %s", pj_get_version(),
+    pj_ansi_snprintf(tmp, sizeof(tmp), "PJSUA v%s %s", pj_get_version(),
                     pj_get_sys_info()->info.ptr);
     pj_strdup2_with_null(app_config.pool, &cfg->cfg.user_agent, tmp);
```
Comment (jimying) [inline pjsip-apps/src/pjsua/pjsua_app_config.c:1606]: snprintf code indent all not align  in this file

File: pjsip/src/pjsua-lib/pjsua_call.c [hunk 15]
```diff
@@ -697,10 +695,11 @@ static pj_status_t apply_call_setting(pjsua_call *call,
     /* If call is established or media channel hasn't been initialized,
      * reinit media channel.
      */
-    if ((call->inv && call->inv->state == PJSIP_INV_STATE_CONNECTING &&
-         call->med_cnt == 0) ||
-        (call->inv && call->inv->state == PJSIP_INV_STATE_CONFIRMED) ||
-        (call->opt.flag & PJSUA_CALL_REINIT_MEDIA))
+    if (call-> inv &&
+        ((call->inv->state == PJSIP_INV_STATE_CONNECTING &&
```
Comment (jimying) [inline pjsip/src/pjsua-lib/pjsua_call.c:699]: `call->  inv`   change to `call->inv`

File: pjsip/src/pjsua-lib/pjsua_acc.c [hunk 16]
```diff
@@ -3519,7 +3519,7 @@ pj_status_t pjsua_acc_get_uac_addr(pjsua_acc_id acc_id,
     {
         int i;
 
-        for (i = 0; i < (int)sizeof(pjsua_var.tpdata); i++) {
+        for (i = 0; i < PJ_ARRAY_SIZE(pjsua_var.tpdata); i++) {
```
Comment (jimying) [inline pjsip/src/pjsua-lib/pjsua_acc.c:3522]: To supress build warning
```
../src/pjsua-lib/pjsua_acc.c:3522:23 warning: 
comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
           for (i = 0; i < PJ_ARRAY_SIZE(pjsua_var.tpdata); i++) {
                     ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

should change to `(int)`PJ_ARRAY_SIZE(pjsua_var.tpdata)

File: pjlib/src/pjlib-test/string.c [hunk 17]
```diff
@@ -289,6 +291,324 @@ static int strcmp_test(void)
...
+    if (exp_dst) {
+        if (strcmp(dst, exp_dst)) {
+            PJ_LOG(3,("", "  strxcpy \"%s\", dst_size=%d: dst content mismatch: \"%s\"!=\"%s\"",  
+                          src, dst_size, dst, exp_dst));
+            return -708;
+        }
+    }
+
+    /* verify not writing pass buffer */
+    for (i=exp_dst?strlen(exp_dst)+1:0; i<sizeof(dst); ++i) {
```
Comment (jimying) [inline pjlib/src/pjlib-test/string.c:324]: To supress build warning
```
./src/pjlib-test/string.c:324:42: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
      for (i=exp_dst?strlen(exp_dst)+1:0; i<sizeof(dst); ++i) {
```

File: pjlib/src/pjlib-test/string.c [hunk 18]
```diff
@@ -289,6 +291,324 @@ static int strcmp_test(void)
...
+    int i, ret;
+
+    PJ_ASSERT_RETURN(src && strlen(cdst) <= 4, -730);
+    PJ_ASSERT_RETURN(strlen(cdst) < dst_size ||
+                     (strlen(cdst)==0 && dst_size==0), -731);
+
+    memset(dst, GUARD, sizeof(dst));
+    if (dst_size) {
+        ret = pj_ansi_strxcpy(dst, cdst, dst_size);
+        PJ_ASSERT_RETURN(ret==strlen(cdst), -732);
```
Comment (jimying) [inline pjlib/src/pjlib-test/string.c:527]: To supress build warning
```
../src/pjlib-test/string.c:527:29: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
          PJ_ASSERT_RETURN(ret==strlen(cdst), -732);
```

File: pjlib/src/pjlib-test/string.c [hunk 19]
```diff
@@ -289,6 +291,324 @@ static int strcmp_test(void)
...
+            PJ_LOG(3,("", "  strxcat \"%s\", \"%s\", dst_size=%d: "
+                          "dst content mismatch: \"%s\"!=\"%s\"",  
+                          cdst, src, dst_size, dst, exp_dst));
+            return -736;
+        }
+    }
+
+    /* verify not writing past buffer */
+    for (i=exp_dst?strlen(exp_dst)+1:0; i<sizeof(dst); ++i) {
+        if (dst[i] != GUARD) {
```
Comment (jimying) [inline pjlib/src/pjlib-test/string.c:551]: To suppress build warning
```
../src/pjlib-test/string.c:550:42: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
     for (i=exp_dst?strlen(exp_dst)+1:0; i<sizeof(dst); ++i) {
```
                                         ~^~~~~~~~~~~~

**Review by jimying — COMMENTED** [review]
hi,  there are serveral build warnings
when CFLAGS has -Wextra flag

my test CFLAGS is "-Wextra -Wno-missing-field-initializers"

File: pjlib/src/pj/os_info.c (hunk 4 above)
Comment (bennylp) [reply pjlib/src/pj/os_info.c:82]: intentional to minimize code change (re: git blame)

File: pjlib-util/src/pjlib-util-test/http_client.c (hunk 5 above)
Comment (bennylp) [reply pjlib-util/src/pjlib-util-test/http_client.c:129]: intentional to minimize code change (re: git blame)

File: pjmedia/src/pjmedia-audiodev/wmme_dev.c (hunk 7 above)
Comment (bennylp) [reply pjmedia/src/pjmedia-audiodev/wmme_dev.c:341]: intentional to minimize code change (re: git blame)

File: pjsip-apps/src/pjsua/pjsua_app_config.c (hunk 14 above)
Comment (bennylp) [reply pjsip-apps/src/pjsua/pjsua_app_config.c:1606]: yes, they're intentional to minimize code change (re: git blame)

File: pjlib/src/pjlib-test/timer.c [hunk 20]
```diff
@@ -334,7 +334,7 @@ static int stress_worker(void *arg)
         idx = pj_rand() % ST_ENTRY_COUNT;
         prev_status = pj_atomic_get(tparam->status[idx]);
         if (task == SCHEDULING) {
-            if (prev_status != 0) continue;
+            //if (prev_status != 0) continue;
```
Comment (bennylp) [inline pjlib/src/pjlib-test/timer.c:337]: @sauwming This seems to cause a lot of `Warning! Rescheduling outstanding entry (0xxxx)` warnings (about 4000 in my comp) to be printed. Is it okay?

**Comment by bennylp** [comment]
> hi, there are serveral build warnings when CFLAGS has -Wextra flag
> 
> my test CFLAGS is "-Wextra -Wno-missing-field-initializers"

It compiles cleanly now with (gcc 11.3): `-Werror -Wextra -Wno-missing-field-initializers -Wno-sign-compare -Wno-unused-parameter -Wno-implicit-fallthrough -Wno-shift-negative-value`

Thanks for the review!

**Comment by jimying** [comment]
> It compiles cleanly now with (gcc 11.3): `-Werror -Wextra -Wno-missing-field-initializers -Wno-sign-compare -Wno-unused-parameter -Wno-implicit-fallthrough -Wno-shift-negative-value`
> 

i think can remove '-Wno-sign-compare ' ^_^
 serveral warnings relate to:
```
../src/pjlib-test/string.c:303:22: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    for (int ig=0; ig<sizeof(GUARDS); ++ig) {
                   ~~^~~~~~~~~~~~~~~
../src/pjlib-test/string.c:407:22: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    for (int ig=0; ig<sizeof(GUARDS); ++ig) {
                   ~~^~~~~~~~~~~~~~~
../src/pjlib-test/string.c:435:27: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
        for (i=dst_size; i<sizeof(dst); ++i) {
                         ~^~~~~~~~~~~~
../src/pjlib-test/string.c:535:22: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    for (int ig=0; ig<sizeof(GUARDS); ++ig) {
                   ~~^~~~~~~~~~~~~~~
../src/pjlib-test/string.c:541:33: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
            PJ_ASSERT_RETURN(ret==strlen(cdst), -732);
                             ~~~^ ~~~~~~~~~~~~
../include/pj/assert.h:60:23: note: expanded from macro 'PJ_ASSERT_RETURN'
                if (!(expr)) { pj_assert(expr); return retval; } \ 
                      ^~~~
../src/pjlib-test/string.c:541:33: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
            PJ_ASSERT_RETURN(ret==strlen(cdst), -732);
                             ~~~^ ~~~~~~~~~~~~
../include/pj/assert.h:60:42: note: expanded from macro 'PJ_ASSERT_RETURN'
                if (!(expr)) { pj_assert(expr); return retval; } \ 
                                         ^~~~
../include/pj/assert.h:48:37: note: expanded from macro 'pj_assert'
#   define pj_assert(expr)   assert(expr)
```

**Comment by bennylp** [comment]
> i think can remove '-Wno-sign-compare ' 

Done, thanks.

**Comment by nanangizz** [comment]
Got this warning on MSVC2005:
```
warning C4002: too many actual parameters for macro 'PJ_PRINT_FUNC_DECOR'
```

**Comment by sauwming** [comment]
```
#if defined(__GNUC__) || defined(__clang__)
#  define PJ_PRINT_FUNC_DECOR(idx) __attribute__((format (printf, idx, idx+1)))
#else
#  define PJ_PRINT_FUNC_DECOR()
#endif
```
perhaps should change the second definition to:
`#  define PJ_PRINT_FUNC_DECOR(idx)`

**Comment by nanangizz** [comment]
> perhaps should change the second definition to: `# define PJ_PRINT_FUNC_DECOR(idx)`

Exactly, actually I have fixed it locally, just not sure if I have to create a new PR just for that :)

**Pull request merged**